File: tui/src/scala/tui/widgets/ParagraphWidget.scala
package tui
package widgets

import tui.internal.saturating._
import tui.internal.reflow.{LineComposer, LineTruncator, WordWrapper}

/** A widget to display some text.
  *
  * @param text
  *   The text to display
  * @param block
  *   A block to wrap the widget in
  * @param style
  *   Widget style
  * @param wrap
  *   How to wrap the text
  * @param scroll
  *   Scroll
  * @param alignment
  *   Alignment of the text
  */
case class ParagraphWidget(
    text: Text,
    style: Style = Style.DEFAULT,
    wrap: Option[ParagraphWidget.Wrap] = None,
    scroll: (Int, Int) = (0, 0),
    alignment: Alignment = Alignment.Left
) extends Widget {

  def render(area: Rect, buf: Buffer): Unit = {
    buf.setStyle(area, this.style)
    if (area.height < 1) {
      return
    }

    val styled: Array[StyledGrapheme] =
      text.lines.flatMap { case Spans(spans: Array[Span]) =>
        spans.flatMap(span => span.styledGraphemes(this.style)) :+ StyledGrapheme(Grapheme("\n"), this.style)
      }

    val line_composer: LineComposer =
      wrap match {
        case Some(ParagraphWidget.Wrap(trim)) => WordWrapper(styled.iterator, area.width, trim)
        case None =>
          val line_composer = LineTruncator(styled.iterator, area.width)
          alignment match {
            case Alignment.Left => line_composer.copy(horizontal_offset = this.scroll._2)
            case _              => line_composer
          }
      }

    var y = 0
    var continue = true
    while (continue)
      line_composer.next_line() match {
        case None => continue = false
        case Some((current_line, current_line_width)) =>
          if (y >= this.scroll._1) {
            var x = ParagraphWidget.getLineOffset(current_line_width, area.width, this.alignment)
            current_line.foreach { case StyledGrapheme(symbol, style) =>
              // If the symbol is empty, the last char which rendered last time will
              // leave on the line. It's a quick fix.
              val newSymbol = if (symbol.str.isEmpty) " " else symbol.str

              buf
                .get(area.left + x, area.top + y - this.scroll._1)
                .setSymbol(newSymbol)
                .setStyle(style)

              x += symbol.width
            }
          }
          y += 1
          if (y >= area.height + this.scroll._1) {
            continue = false
          }
      }
  }
}
object ParagraphWidget {
  def getLineOffset(lineWidth: Int, textAreaWidth: Int, alignment: Alignment): Int =
    alignment match {
      case Alignment.Center => (textAreaWidth / 2).saturating_sub_unsigned(lineWidth / 2)
      case Alignment.Right  => textAreaWidth.saturating_sub_unsigned(lineWidth)
      case Alignment.Left   => 0
    }

  /** Describes how to wrap text across lines.
    *
    * @param trim
    *   Should leading whitespace be trimmed
    */
  case class Wrap(
      trim: Boolean
  )
}

File: tui/src/java/tui/widgets/ParagraphWidget.java
package tui.widgets;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Optional;
import tui.Alignment;
import tui.Buffer;
import tui.Grapheme;
import tui.Rect;
import tui.Span;
import tui.Spans;
import tui.Style;
import tui.StyledGrapheme;
import tui.Widget;
import tui.internal.Reflow;
import tui.internal.Saturating;

/// A widget to display some text.
public final class ParagraphWidget implements Widget {
  public final tui.Text text;
  public final Optional<BlockWidget> block;
  public final Style style;
  public final Optional<Wrap> wrap;
  public final Scroll scroll;
  public final Alignment alignment;

  public ParagraphWidget(
      tui.Text text,
      Optional<BlockWidget> block,
      Style style,
      Optional<Wrap> wrap,
      Scroll scroll,
      Alignment alignment) {
    this.text = text;
    this.block = block;
    this.style = style;
    this.wrap = wrap;
    this.scroll = scroll;
    this.alignment = alignment;
  }

  public static ParagraphWidget empty(tui.Text text) {
    return new ParagraphWidget(
        text, Optional.empty(), Style.DEFAULT, Optional.empty(), new Scroll(0, 0), Alignment.Left);
  }

  public ParagraphWidget withBlock(BlockWidget b) {
    return new ParagraphWidget(text, Optional.of(b), style, wrap, scroll, alignment);
  }

  public ParagraphWidget withStyle(Style s) {
    return new ParagraphWidget(text, block, s, wrap, scroll, alignment);
  }

  public ParagraphWidget withWrap(Wrap w) {
    return new ParagraphWidget(text, block, style, Optional.of(w), scroll, alignment);
  }

  public ParagraphWidget withScroll(Scroll s) {
    return new ParagraphWidget(text, block, style, wrap, s, alignment);
  }

  public ParagraphWidget withAlignment(Alignment a) {
    return new ParagraphWidget(text, block, style, wrap, scroll, a);
  }

  @Override
  public void render(Rect area, Buffer buf) {
    buf.setStyle(area, style);
    Rect textArea;
    if (block.isPresent()) {
      BlockWidget b = block.get();
      Rect innerArea = b.inner(area);
      b.render(area, buf);
      textArea = innerArea;
    } else {
      textArea = area;
    }

    if (textArea.height() < 1) {
      return;
    }

    List<StyledGrapheme> styled = new ArrayList<>();
    for (Spans line : text.lines()) {
      for (Span span : line.spans()) {
        styled.addAll(Arrays.asList(span.styledGraphemes(style)));
      }
      styled.add(new StyledGrapheme(new Grapheme("\n"), style));
    }
    Iterator<StyledGrapheme> iter = styled.iterator();

    Reflow.LineComposer lineComposer;
    if (wrap.isPresent()) {
      lineComposer = new Reflow.WordWrapper(iter, textArea.width(), wrap.get().trim());
    } else {
      int horizontalOffset = alignment == Alignment.Left ? scroll.x() : 0;
      lineComposer = new Reflow.LineTruncator(iter, textArea.width(), horizontalOffset);
    }

    int y = 0;
    boolean cont = true;
    while (cont) {
      Optional<Reflow.Line> nextLine = lineComposer.nextLine();
      if (nextLine.isEmpty()) {
        cont = false;
      } else {
        Reflow.Line line = nextLine.get();
        if (y >= scroll.y()) {
          int x = getLineOffset(line.width(), textArea.width(), alignment);
          for (StyledGrapheme sg : line.graphemes()) {
            String newSymbol = sg.symbol().str.isEmpty() ? " " : sg.symbol().str;
            buf.get(textArea.left() + x, textArea.top() + y - scroll.y())
                .setSymbol(newSymbol)
                .setStyle(sg.style());
            x += sg.symbol().width();
          }
        }
        y += 1;
        if (y >= textArea.height() + scroll.y()) {
          cont = false;
        }
      }
    }
  }

  public static int getLineOffset(int lineWidth, int textAreaWidth, Alignment alignment) {
    return switch (alignment) {
      case Center -> Saturating.saturatingSubUnsigned(textAreaWidth / 2, lineWidth / 2);
      case Right -> Saturating.saturatingSubUnsigned(textAreaWidth, lineWidth);
      case Left -> 0;
    };
  }

  /// Describes how to wrap text across lines.
  public record Wrap(boolean trim) {}

  public record Scroll(int y, int x) {}
}
